### ISSUE TITLE
SMW\Exception\DataItemException from line 26 of SMW_DI_Number.php: Initialization value '' is not a number.

### ISSUE THREAD

**Issue by kghbln** [SemanticMediaWiki/SemanticExtraSpecialProperties#77]
### Setup and configuration

- MediaWiki 	1.29.0-rc.0 (c749bf6) 14:52, 25 May 2017
- PHP 	5.6.30-0+deb8u1 (apache2handler)
- MariaDB 	10.0.31-MariaDB-1~jessie
- Semantic Extra Special Properties	2.0.0-alpha (5ce8bdc) 13:10, 4 June 2017	GPL-2.0+

### Issue

```
(1280/14307)    Finished processing ID 1295* (Attribut:Polarity)
[817f9bdd9e822151658259b5] [no req]   SMW\Exception\DataItemException from line 26 of /../w/extensions/SemanticMediaWiki/includes/dataitems/SMW_DI_Number.php: Initialization value '' is not a number.
```
### Backtrace

```
#0 /../w/extensions/SemanticExtraSpecialProperties/src/PropertyAnnotators/UserEditCountPropertyAnnotator.php(69): SMWDINumber->__construct(NULL)
#1 /../w/extensions/SemanticExtraSpecialProperties/src/ExtraPropertyAnnotator.php(90): SESP\PropertyAnnotators\UserEditCountPropertyAnnotator->addAnnotation(SMW\DIProperty, SMW\SemanticData)
#2 /../w/extensions/SemanticExtraSpecialProperties/src/HookRegistry.php(159): SESP\ExtraPropertyAnnotator->addAnnotation(SMW\SemanticData)
#3 [internal function]: SESP\HookRegistry->SESP\{closure}(SMWSQLStore3, SMW\SemanticData)
#4 /../w/includes/Hooks.php(186): call_user_func_array(Closure, array)
#5 /../w/extensions/SemanticMediaWiki/includes/storage/SMW_Store.php(218): Hooks::run(string, array)
#6 /../w/extensions/SemanticMediaWiki/includes/StoreUpdater.php(218): SMW\Store->updateData(SMW\SemanticData)
#7 /../w/extensions/SemanticMediaWiki/includes/StoreUpdater.php(142): SMW\StoreUpdater->doRealUpdate()
#8 /../w/extensions/SemanticMediaWiki/includes/StoreUpdater.php(96): SMW\StoreUpdater->doPerformUpdate()
#9 /../w/extensions/SemanticMediaWiki/includes/ParserData.php(406): SMW\StoreUpdater->doUpdate()
#10 [internal function]: SMW\ParserData->SMW\{closure}()
#11 /../w/extensions/SemanticMediaWiki/src/Updater/DeferredCallableUpdate.php(236): call_user_func(Closure)
#12 /../w/extensions/SemanticMediaWiki/src/Updater/TransactionalDeferredCallableUpdate.php(134): SMW\Updater\DeferredCallableUpdate->doUpdate()
#13 /../w/extensions/SemanticMediaWiki/src/Updater/DeferredCallableUpdate.php(265): SMW\Updater\TransactionalDeferredCallableUpdate->doUpdate()
#14 /../w/extensions/SemanticMediaWiki/includes/ParserData.php(422): SMW\Updater\DeferredCallableUpdate->pushUpdate()
#15 /../w/extensions/SemanticMediaWiki/src/MediaWiki/Jobs/UpdateJob.php(184): SMW\ParserData->updateStore()
#16 /../w/extensions/SemanticMediaWiki/src/MediaWiki/Jobs/UpdateJob.php(146): SMW\MediaWiki\Jobs\UpdateJob->updateStore(SMW\ParserData)
#17 /../w/extensions/SemanticMediaWiki/src/MediaWiki/Jobs/UpdateJob.php(118): SMW\MediaWiki\Jobs\UpdateJob->needToParsePageContentBeforeUpdate()
#18 /../w/extensions/SemanticMediaWiki/src/MediaWiki/Jobs/UpdateJob.php(86): SMW\MediaWiki\Jobs\UpdateJob->doPrepareForUpdate()
#19 /../w/extensions/SemanticMediaWiki/src/MediaWiki/Jobs/UpdateJob.php(67): SMW\MediaWiki\Jobs\UpdateJob->doUpdate()
#20 /../w/extensions/SemanticMediaWiki/src/SQLStore/EntityRebuildDispatcher.php(198): SMW\MediaWiki\Jobs\UpdateJob->run()
#21 /../w/extensions/SemanticMediaWiki/src/Maintenance/DataRebuilder.php(307): SMW\SQLStore\EntityRebuildDispatcher->startRebuildWith(integer)
#22 /../w/extensions/SemanticMediaWiki/src/Maintenance/DataRebuilder.php(266): SMW\Maintenance\DataRebuilder->doExecuteFor(SMW\SQLStore\EntityRebuildDispatcher, integer)
#23 /../w/extensions/SemanticMediaWiki/src/Maintenance/DataRebuilder.php(168): SMW\Maintenance\DataRebuilder->doRebuildAll()
#24 /../w/extensions/SemanticMediaWiki/maintenance/rebuildData.php(150): SMW\Maintenance\DataRebuilder->rebuild()
#25 /../w/maintenance/doMaintenance.php(111): SMW\Maintenance\RebuildData->execute()
#26 /../w/extensions/SemanticMediaWiki/maintenance/rebuildData.php(198): require_once(string)
#27 {main}
```
#### ID 1295
```json
{
    "1295": {
        "smw_id": "1295",
        "smw_title": "Earth",
        "smw_namespace": "0",
        "smw_iw": "",
        "smw_subobject": "",
        "smw_sortkey": "Earth"
    }
}
```
```json
{
    "1295": {
        "smw_di_wikipage": {
            "s_id": "1295"
        },
        "smw_fpt_mdat": {
            "s_id": "1295"
        },
        "smw_fpt_cdat": {
            "s_id": "1295"
        },
        "smw_fpt_newp": {
            "s_id": "1295"
        },
        "smw_fpt_ledt": {
            "s_id": "1295"
        },
        "smw_ftp_sesp_revid": {
            "s_id": "1295"
        },
        "smw_ftp_sesp_pageid": {
            "s_id": "1295"
        },
        "smw_ftp_sesp_pagelgth": {
            "s_id": "1295"
        },
        "smw_ftp_sesp_euser": {
            "s_id": "1295"
        },
        "smw_ftp_sesp_cuser": {
            "s_id": "1295"
        },
        "smw_ftp_sesp_views": {
            "s_id": "1295"
        },
        "smw_ftp_sesp_nrev": {
            "s_id": "1295"
        }
    }
}
```

#### ID 1296

```json
{
    "1296": {
        "smw_id": "1296",
        "smw_title": "Erlangen",
        "smw_namespace": "0",
        "smw_iw": "",
        "smw_subobject": "",
        "smw_sortkey": "Erlangen"
    }
}
```
```json
{
    "1296": {
        "smw_di_wikipage": {
            "s_id": "1296"
        },
        "smw_fpt_mdat": {
            "s_id": "1296"
        },
        "smw_fpt_cdat": {
            "s_id": "1296"
        },
        "smw_fpt_newp": {
            "s_id": "1296"
        },
        "smw_fpt_ledt": {
            "s_id": "1296"
        },
        "smw_ftp_sesp_revid": {
            "s_id": "1296"
        },
        "smw_ftp_sesp_pageid": {
            "s_id": "1296"
        },
        "smw_ftp_sesp_pagelgth": {
            "s_id": "1296"
        },
        "smw_ftp_sesp_euser": {
            "s_id": "1296"
        },
        "smw_ftp_sesp_cuser": {
            "s_id": "1296"
        },
        "smw_ftp_sesp_views": {
            "s_id": "1296"
        },
        "smw_ftp_sesp_nrev": {
            "s_id": "1296"
        }
    }
}
```
However [Erlangen](https://sandbox.semantic-mediawiki.org/wiki/Sp%C3%A9cial:Parcourir/:Erlangen) seems to be ok.

**Comment by kghbln** [comment]
This seems to be a bigger problem than expected. The server just crashed because `rebuildData.php` produced a exception log file of 45 GB which filled up the remaining disk space.

Disposing the IDs 1295 and 1296 did not help either since the script continuous to fatal on this spot.

**Comment by mwjames** [comment]
Hi,

> Disposing the IDs 1295 and 1296 did not help either since the script
> continuous to fatal on this spot.

As quick fix:


```
-		if ( $user instanceof User ) {
-			$dataItem = new DINumber( $user->getEditCount() );
+		if ( $user instanceof User && ( $count = $user->getEditCount() ) &&
is_int( $count ) ) {
+			$dataItem = new DINumber( $count );
 		}
```

I'll post fix during the weekend.

Cheers

**Comment by mwjames** [comment]
> This seems to be a bigger problem than expected. The server just crashed
> because `rebuildData.php` produced a exception log file of 45 GB which
> filled up the remaining disk space.

Let's crash the server with logs.

**Comment by kghbln** [comment]
Good that I added `_VIEWS` to the array of used special properties a couple of days ago.

> As quick fix:

This is [L68ff](https://github.com/SemanticMediaWiki/SemanticExtraSpecialProperties/blob/5ce8bdc2740f4090abc503acfda8d2f449dc33d9/src/PropertyAnnotators/UserEditCountPropertyAnnotator.php#L68) and so I locally adapted.

"rebuildData.php" ran smooth now and the prompt revealed a [redirect](https://sandbox.semantic-mediawiki.org/w/index.php?title=Utilisateur:Hans_Oleander_sandbox&redirect=no) as the cause of the issue.

I was not able to preserve the exception log but the smwdebug log (only 1.4 GB) reveals the looping:

```
2017-06-06 20:12:49 phalaris 0210020150926-02100_: SMW\Protection\EditProtectionUpdater::doUpdateFrom no update required
2017-06-06 20:12:49 phalaris 0210020150926-02100_: Event: cached.propertyvalues.prefetcher.reset :: Hans_Oleander_sandbox#2#
```
I have about 1.6 mio lines of this it that file. It is safe to say that the accompanying exception log held the exception originally posted times that number.

> Let's crash the server with logs.

This happened fast. Usually I get a 90% disc space warning (= only 20 GB left). Since I did not get one yesterday. Apart from that. That's why I provide this test instance only hosting wikis which may be unavailable for a couple of hours without creating pain.
